Python/codecs.c

```diff
@@ -864,108 +864,105 @@ PyObject *PyCodec_XMLCharRefReplaceErrors(PyObject *exc)
     if (PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeDecodeError)) {
-        const unsigned char *p;
-        if (PyUnicodeDecodeError_GetStart(exc, &start))
-            return NULL;
-        if (PyUnicodeDecodeError_GetEnd(exc, &end))
-            return NULL;
-        if (!(object = PyUnicodeDecodeError_GetObject(exc)))
+        if (_PyUnicodeError_GetParams(exc,
+                                      &obj, &objlen,
+                                      &start, &end, &slen, true) < 0)
+        {
             return NULL;
-        p = (const unsigned char*)PyBytes_AS_STRING(object);
-        res = PyUnicode_New(4 * (end - start), 127);
+        }
+        PyObject *res = PyUnicode_New(4 * slen, 127);
         if (res == NULL) {
-            Py_DECREF(object);
+            Py_DECREF(obj);
             return NULL;
         }
-        outp = PyUnicode_1BYTE_DATA(res);
-        for (i = start; i < end; i++, outp += 4) {
-            unsigned char c = p[i];
+        Py_UCS1 *outp = PyUnicode_1BYTE_DATA(res);
+        const unsigned char *p = (const unsigned char *)PyBytes_AS_STRING(obj);
+        for (Py_ssize_t i = start; i < end; i++, outp += 4) {
+            const unsigned char ch = p[i];
             outp[0] = '\\';
             outp[1] = 'x';
-            outp[2] = Py_hexdigits[(c>>4)&0xf];
-            outp[3] = Py_hexdigits[c&0xf];
+            outp[2] = Py_hexdigits[(ch >> 4) & 0xf];
+            outp[3] = Py_hexdigits[ch & 0xf];
         }
-
         assert(_PyUnicode_CheckConsistency(res, 1));
-        Py_DECREF(object);
+        Py_DECREF(obj);
         return Py_BuildValue("(Nn)", res, end);
     }
-    if (PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeEncodeError)) {
-        if (PyUnicodeEncodeError_GetStart(exc, &start))
-            return NULL;
-        if (PyUnicodeEncodeError_GetEnd(exc, &end))
-            return NULL;
-        if (!(object = PyUnicodeEncodeError_GetObject(exc)))
-            return NULL;
-    }
-    else if (PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeTranslateError)) {
-        if (PyUnicodeTranslateError_GetStart(exc, &start))
-            return NULL;
-        if (PyUnicodeTranslateError_GetEnd(exc, &end))
-            return NULL;
-        if (!(object = PyUnicodeTranslateError_GetObject(exc)))
+
+    if (
+        PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeEncodeError)
+        || PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeTranslateError)
+    ) {
+        if (_PyUnicodeError_GetParams(exc,
+                                      &obj, &objlen,
+                                      &start, &end, &slen, false) < 0)
+        {
             return NULL;
+        }
     }
     else {
         wrong_exception_type(exc);
         return NULL;
     }
 
-    if (end - start > PY_SSIZE_T_MAX / (1+1+8))
-        end = start + PY_SSIZE_T_MAX / (1+1+8);
-    for (i = start, ressize = 0; i < end; ++i) {
+    if (slen == 0) { // end <= start
```

Comment:
Is it needed to test explicitly for slen==0? ressize will be zero anyway, no?

Reply:
It's a small optimization but I can remove it if you want (it just avoids having two empty loops, an optimized call to `PyUnicode_New` (since slen = 0), and a call to `PyUnicode_1BYTE_DATA`).

It would be a bit cleaner without the optimization though. Maybe it's indeed better to remove it.

Reply:
I suggest to remove it. I don't think that it's common to call an error handler to replace nothing. It's not a performance critical case.

Reply:
I've removed it yes. It makes the code cleaner and maybe the compiler is smart.